File: suites/osprey-compute-kernels/reference/cases/nestedloop/nestedloop.hs
-- Triple-nested counting loop accumulating (i*j*k) mod P — nested iteration.
modP :: Int
modP = 1000000007

loopK :: Int -> Int -> Int -> Int -> Int
loopK i j k acc
  | k == 0 = acc
  | otherwise = loopK i j (k - 1) ((acc + i * j * k) `mod` modP)

loopJ :: Int -> Int -> Int -> Int -> Int
loopJ i j n acc
  | j == 0 = acc
  | otherwise = loopJ i (j - 1) n (loopK i j n acc)

loopI :: Int -> Int -> Int -> Int
loopI i n acc
  | i == 0 = acc
  | otherwise = loopI (i - 1) n (loopJ i n n acc)

main :: IO ()
main = print (loopI 250 250 0)

File: suites/osprey-compute-kernels/reference/cases/nestedloop/nestedloop.ml
(* Triple-nested counting loop accumulating (i*j*k) mod P — nested iteration. *)
let modp = 1000000007

let rec loop_k i j k acc =
  if k = 0 then acc
  else loop_k i j (k - 1) ((acc + i * j * k) mod modp)

let rec loop_j i j n acc =
  if j = 0 then acc
  else loop_j i (j - 1) n (loop_k i j n acc)

let rec loop_i i n acc =
  if i = 0 then acc
  else loop_i (i - 1) n (loop_j i n n acc)

let () = Printf.printf "%d\n" (loop_i 250 250 0)
